packages/NodeCollector/BinaryOpTreeRootLocator.php

```diff
@@ -25,31 +25,27 @@ final class BinaryOpTreeRootLocator
     public function findOperationRoot(Expr $expr, string $binaryOpClass): Expr
     {
         $parentNode = $expr->getAttribute(AttributeKey::PARENT_NODE);
-        if (! $parentNode instanceof Node) {
+        if ($parentNode === null) {
+            // No more parents so the Expr node must be root.
             return $expr;
         }
 
         if ($parentNode::class !== $binaryOpClass) {
+            // If the Expr node is not a child of the desired operation,
+            // it must already be the root of the operation tree.
             return $expr;
         }
 
-        if (! $parentNode instanceof BinaryOp) {
-            return $expr;
-        }
-
-        if ($parentNode->left === $expr) {
-            return $this->findOperationRoot($parentNode, $binaryOpClass);
-        }
-
-        $isRightChild = $parentNode->right === $expr;
-        if (! $isRightChild) {
-            return $expr;
-        }
-
-        if ($expr::class === $binaryOpClass) {
+        $isParentARightAssociativeTree = $parentNode->right === $expr && $expr::class === $binaryOpClass;
+        if ($isParentARightAssociativeTree) {
```

Comment:
Here I factored out the condition or `ChangeAndIfToEarlyReturnRector` would botch it up again.